src/System.Private.ServiceModel/src/System/ServiceModel/Channels/CommunicationObject.cs

```diff
@@ -808,7 +851,7 @@ public static void OnClose(CommunicationObject communicationObject, TimeSpan tim
 
         public static IAsyncResult OnBeginClose(CommunicationObject communicationObject, TimeSpan timeout, AsyncCallback callback, object state)
         {
-            return communicationObject.OnCloseAsync(timeout).ToApm(callback, state);
+            return communicationObject.CallOnCloseAsync(timeout).ToApm(callback, state);
```

Comment:
Could end up in an infinite loop if the caller doesn't implement the APM methods either. Probably want to only call CallOnCloseAsync if _canUseIAsyncCommunicationObject otherwise have the call terminating with Task.Completed.ToApm(callback, state); The same on the other calls.